tools/walletextension/obscuro_facade.go

```diff
@@ -0,0 +1,119 @@
+package walletextension
+
+import (
+	"crypto/ecdsa"
+	"crypto/rand"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/ethereum/go-ethereum/crypto/ecies"
+	"github.com/gorilla/websocket"
+)
+
+// ObscuroFacade is a server that inverts the encryption and decryption performed by WalletExtension, so that the forwarded
+// Ethereum JSON-RPC requests can be understood by a regular Geth node.
+type ObscuroFacade struct {
+	enclavePrivateKey *ecdsa.PrivateKey
+	gethWebsocketAddr string
+	viewingKeyChannel <-chan ViewingKey
+	viewingKey        *ecdsa.PublicKey
+}
+
+func NewObscuroFacade(
+	enclavePrivateKey *ecdsa.PrivateKey,
+	gethWebsocketAddr string,
+	viewingKeyChannel <-chan ViewingKey,
+) *ObscuroFacade {
+	return &ObscuroFacade{
+		enclavePrivateKey: enclavePrivateKey,
+		gethWebsocketAddr: gethWebsocketAddr,
+		viewingKeyChannel: viewingKeyChannel,
+	}
+}
+
+// Serve listens for and serves Ethereum JSON-RPC requests from the wallet extension.
+func (of *ObscuroFacade) Serve(hostAndPort string) {
+	// We listen for the account viewing key.
+	go func() {
+		viewingKey := <-of.viewingKeyChannel
+		// TODO - Verify signed bytes.
+		of.viewingKey = viewingKey.viewingKeyPublic
+	}()
+
+	serveMux := http.NewServeMux()
+
+	serveMux.HandleFunc(pathRoot, of.handleWSEthJSON)
+
+	err := http.ListenAndServe(hostAndPort, serveMux)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (of *ObscuroFacade) handleWSEthJSON(resp http.ResponseWriter, req *http.Request) {
+	upgrader := websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+	}
+	connection, err := upgrader.Upgrade(resp, req, nil)
+	if err != nil {
+		http.Error(resp, fmt.Sprintf("could not upgrade connection to a websocket connection: %v", err), httpCodeErr)
+		return
+	}
+
+	// We read the message from the wallet extension.
+	_, encryptedMessage, err := connection.ReadMessage()
+	if err != nil {
+		msg := fmt.Sprintf("could not read Ethereum JSON-RPC request: %v", err)
+		_ = connection.WriteMessage(websocket.TextMessage, []byte(msg))
+		return
+	}
```

Comment:
Should these errors be logged ?

Reply:
I should be logging them. I think I'll log them on the wallet extension side only, to not double-log them (once on the facade, then again on the wallet extension).